architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,652 @@
+  Resources []BicepDeploymentResource `json:"resources,omitempty"`
+}
+```
+
+#### `rad bicep generate-kubernetes` Command
+
+The `rad bicep generate-kubernetes` command will be a new command that will generate a Kubernetes custom resource when provided a Bicep template, modules, and parameters The command will take the following arguments:
+
+- `-p, --parameters`: The parameters for the Bicep template. Can be specified in the same way as [`rad deploy` parameters](https://edge.docs.radapp.io/reference/cli/rad_deploy/).
+- `-o, --outfile`: The path to the output file where the BicepDeployment resource will be written.
+
+#### BicepDeployment Controller
+
+The `radius-controller` will be updated to include a new controller that reconciles BicepDeployment resources. It will have the following control loop:
+
+ The BicepDeployment controller will adhere to the following control loop for BicepDeployment resources on the cluster:
+
+1. Check if there is an in-progress operation. If so, check its status:
+    1. If the operation is still in progress, then queue another reconcile operation and continue processing.
+	  2. If the operation completed successfully, then update the `status.phrase` and `status.description` fields as succeeded. Set the contents of the `status.resources` to match the set of deployed resources. This operation is complete, so we can continue processing.
+	  3. If the operation failed, then update the `status.phrase` and `status.description` as failed and continue processing.
+2. If the BicepDeployment is being deleted, then process deletion:
+    1. Queue a DELETE operation for each of the resources in the `status.resources` list.
+    2. After all resources are deleted, update the `status.phrase` and `status.description` fields as deleted.
+3. If the BicepDeployment is not being deleted then process this as a create or update:
+    1. Compile the Bicep template, modules, and parameters into ARM JSON.
+    1. The controller will perform some simple lookups to Radius to determine if the provided template can be deployed:
+        - Check if the template contains an empty `param environment`. If it does, the controller will query the Radius API to get an environment ID (must have exactly one environment if left unspecified like this) and deploy the template using that environment.
+        - Check if the template contains an empty `param application`. If it does, the controller will query the Radius API to get an application ID. If there is exactly one application present, the controller will deploy the template using that application. If there are no applications, then the controller will create one.
+    2. Queue a PUT operation against the Radius API to deploy compiled ARM JSON.
+
+The controller will also be shipped with a Bicep binary and a `bicepconfig.json` which will allow it to compile Bicep templates and parameters into ARM JSON. The Bicep binary will be included in the `radius-controller` container image.
```

Comment:
https://github.com/Azure/bicep/issues/12290 is something you might be interested in.